realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -907,66 +959,61 @@ void BaseRealSenseNode::publishFrame(
         height = timage.get_height();
         bpp = timage.get_bytes_per_pixel();
     }
-    auto& image = images[stream];
 
-    if (image.size() != cv::Size(width, height) || image.depth() != _image_formats[bpp])
+    if (image_publishers.find(stream) != image_publishers.end())
     {
-        image.create(height, width, _image_formats[bpp]);
-    }
-    image.data = (uint8_t*)f.get_data();
+        auto &image_publisher = image_publishers.at(stream);
+        cv::Mat image_cv_matrix;
+        if (_rgbd_publisher && 0 != _rgbd_publisher->get_subscription_count())
+        {
+            image_cv_matrix = getCVMatImage(f, images, width, height, bpp, stream);
+        }
+        if (0 != image_publisher->get_subscription_count())
+        {
+            if(image_cv_matrix.empty())
+            {
+                image_cv_matrix = getCVMatImage(f, images, width, height, bpp, stream);
+            }
 
-    if (f.is<rs2::depth_frame>())
-    {
-        image = fix_depth_scale(image, _depth_scaled_image[stream]);
-    }
+            // Prepare image topic to be published
+            // We use UniquePtr for allow intra-process publish when subscribers of that type are available
+            sensor_msgs::msg::Image::UniquePtr img_msg_ptr(new sensor_msgs::msg::Image());
+            fillMessageImage(image_cv_matrix, stream, width, height, bpp, t, img_msg_ptr.get());
+            if (!img_msg_ptr)
+            {
+                ROS_ERROR("sensor image message allocation failed, frame was dropped");
+                return;
+            }
+            
+            // Transfer the unique pointer ownership to the RMW
+            sensor_msgs::msg::Image *msg_address = img_msg_ptr.get();
+            image_publisher->publish(std::move(img_msg_ptr));
 
+            ROS_DEBUG_STREAM(rs2_stream_to_string(f.get_profile().stream_type()) << " stream published, message address: " << std::hex << msg_address);
+        }
+    }
     // if stream is on, and is not a SC stream, publish cameraInfo
```

Comment:
This code needs dome refactoring IMO..
can we try splitting the camera info publish to a function?
Something like
```
if (should_publish_camera_info())
{
    publish_camera_info(...);
}
```

Reply:
I prefer to do this refactoring later, with a bigger redesign to all topics/publishers architecture, after finishing this high priority feature.

Reply:
NP